content/posts/expose-serial-port.md

```diff
@@ -0,0 +1,190 @@
++    "Path to a serial device that should be attached to crosvm "
++    "(/dev/<something>). "
++    "QEMU is not supported.\n"
++    "To see the attached device name inside the guest machine, look for a "
++    "corresponding log "
++    "message with the 'SERIAL_PORT' prefix during the VM start.\n"
++    "The device name is stable, but it may change on AOSP updates.");
++
+ DEFINE_vec(start_webrtc, cuttlefish::BoolToString(CF_DEFAULTS_START_WEBRTC),
+             "Whether to start the webrtc process.");
+ 
+@@ -818,6 +830,10 @@ Result<CuttlefishConfig> InitializeCuttlefishConfiguration(
+ 
+   tmp_config_obj.set_gem5_debug_flags(FLAGS_gem5_debug_flags);
+ 
++  tmp_config_obj.set_attach_serial_device(FLAGS_attach_serial_device);
++
+   // streaming, webrtc setup
+   tmp_config_obj.set_webrtc_certs_dir(FLAGS_webrtc_certs_dir);
+   tmp_config_obj.set_sig_server_secure(FLAGS_webrtc_sig_server_secure);
+```
+
+Cuttlefish has already quite a few defined serial ports, let's increase their number by one:
+```
+diff --git a/host/libs/vm_manager/vm_manager.h b/host/libs/vm_manager/vm_manager.h
+index 4b116bfd9..fa300a213 100644
+--- a/host/libs/vm_manager/vm_manager.h
++++ b/host/libs/vm_manager/vm_manager.h
+@@ -54,7 +54,8 @@ class VmManager {
+   // - /dev/hvc9 = uwb
+   // - /dev/hvc10 = oemlock
+   // - /dev/hvc11 = keymint
+-  static const int kDefaultNumHvcs = 12;
++  static const int kDefaultNumHvcs =
++    13; // NOTE: on merge conflicts do +1 to the upstream's value
+ 
+   // This is the number of virtual disks (block devices) that should be
+   // configured by the VmManager. Related to the description above regarding
+```
+
+And finally let's actually forward the new parameter to `crosvm`:
+```
+diff --git a/host/libs/vm_manager/crosvm_manager.cpp b/host/libs/vm_manager/crosvm_manager.cpp
+index 100b71e56..3f568ce28 100644
+--- a/host/libs/vm_manager/crosvm_manager.cpp
++++ b/host/libs/vm_manager/crosvm_manager.cpp
+@@ -449,6 +449,20 @@ Result<std::vector<MonitorCommand>> CrosvmManager::StartCommands(
+   for (auto i = 0; i < VmManager::kMaxDisks - disk_num; i++) {
+     crosvm_cmd.AddHvcSink();
+   }
++
++  if (!config.attach_serial_device().empty()) {
++    crosvm_cmd.AddHvcReadWrite(config.attach_serial_device(),
++                               config.attach_serial_device());
++    LOG(INFO) << "SERIAL_PORT: attaching serial port to crosvm: "
++              << config.attach_serial_device()
++              << ", most likely device name inside the guest machine: /dev/hvc"
++              << crosvm_cmd.HvcNum() - 1;
++  } else {
++    crosvm_cmd.AddHvcSink();
++  }
++
+   CF_EXPECT(crosvm_cmd.HvcNum() + disk_num ==
+                 VmManager::kMaxDisks + VmManager::kDefaultNumHvcs,
+             "HVC count (" << crosvm_cmd.HvcNum() << ") + disk count ("
+```
+
+... but not to QEMU, which is mostly used as AVD for normal (phone) Android apps (it might actually work, but we couldn't test because of unrelated problems):
```

Comment:
Couldn't test it on QEMU unfortunately - it's broken on Sterling at the moment.

But I could of course clone Google's AOSP and try this change there to see if it works on QEMU.

Do you think I should do it? Or does Trout suffice?

Reply:
As the title mentions trout/cuttlefish I don't think we need to do anything for QEMU. What I would do is rephrase the end of the line: it might actually work, but we haven't test it since our focus is Trout/Crossvm.